content/article/my-perl-wishlist-invariant-sigils-part-1.md

```diff
@@ -0,0 +1,181 @@
+    "authors"     : ["christopher-white"],
+    "date"        : "2019-10-15T08:33:34",
+    "tags"        : ["perl-internals", "syntax", "internals",
+                    "compiler", "perl-5", "language", "language-design",
+                    "sigil", "parser", "parsing", "perl-parsing", "core", "c"],
+    "draft"       : true,
+    "image"       : "",
+    "thumbnail"   : "",
+    "description" : "The change I would most like to see in Perl 5, and why it might not happen",
+    "categories"  : "programming-languages"
+  }
+
+Pop quiz!  Q: What was my mistake in this line?
+
+```perl
+is(%HASH{answer}, 'forty-two', q[%HASH properly filled]);
+```
+
+A: I had the answer right, but I messed up the sigil on `HASH`.  It
+should be:
+
+```perl
+is($HASH{answer}, 'forty-two', q[%HASH properly filled]);
+#  ^ $, not %
+```
+
+Unfortunately, on Perl 5.20+, both statements work the same way!  I
+didn't catch the problem until I shipped this code and
+[cpantesters](http://matrix.cpantesters.org/?dist=vars-i+1.08-TRIAL)
+showed me my mistake.  It was an easy fix, but it reminded me that Perl's
+[variant sigils](http://modernperlbooks.com/books/modern_perl/chapter_03.html#variablenamesandsigils)
+can trip up programmers at any level.  If I could change one thing about
+Perl 5, I would change to invariant sigils.
+
+The current situation
+---------------------
+
+In Perl, the sigil tells you
+[how many things to expect](https://perldoc.perl.org/5.30.0/perlintro.html#Perl-variable-types).
+Scalars such as `$foo` are single values.  Any single value in an array
+`@foo` or hash `%foo`, since it is only one thing,
+[also uses `$`](https://perldoc.perl.org/5.30.0/perldata.html#Variable-names),
+so `$foo`, `@foo`, and `%foo` could all refer to different pieces of the
+same variable.  This technique of "variant sigils" works, but confuses
+new Perl users and tripped up yours truly.  To know what you
+are accessing in an array or hash, you have to look at both the sigil
+and the brackets.  As a reminder:
+
+| Sigil | No brackets                                     | `[ ]` (array access)                                                                              | `{ }` (hash access)                                                             |
+|-------|-------------------------------------------------|---------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------|
+| `$`   | `$z`: a scalar, i.e., a single value            | `$z[0]`: the first element of `@z`                                                                | `$z{0}`: the value in `%z` at key `"0"`                                         |
+| `@`   | `@z`: An array, i.e., a list of value(s)        | `@z[0]`: a list of one element from `@z` (an "array slice")                                       | `@z{0}`: the list containing the value `$z{0}` from hash `%z`                   |
```

Comment:
I like how you've examined the different syntax examples. One thing that could make the slices easier to grok is taking multiple elements in the slice. E.g:

    `@z[0, 1]`

Which makes them more distinct from the single element access syntax.

Reply:
Please see the new version once I update the PR and let me know if it works better.